### ISSUE TITLE
[Link Hander] expose interface of Link Handler to be able to use it in other Models

### ISSUE THREAD

**Issue by jkoebner** [adobe/aem-core-wcm-components#1818]
## Link Handler should be exposed as interface so i can use it when including Core Components as dependency

**The issue**
When trying to reuse the models provided by the core components in a project that includes the core components as dependency i can neither use the NavigationItem nor the AbstractLinkItems nor the LinkHandler. The different parts that can be adapted as sling models should be exposed so i can reuse them and so i do not have to copy paste the code from the repo

**Solution**
An Interface for LinkHandler should be added and the current implementation in the internal package should implement the interface so other modules can use that LinkHandler

**Comment by ky940819** [comment]
I was planning on making an issue in PR for this exact feature!

I feel that this is an important issue because I have several custom model implementations that use the Core Components model interfaces; however, without access to the `LinkHandler` it is nearly impossible to properly implement the interfaces without rewriting the entire `LinkHandler`. 

However, the `LinkHandler` methods should not provide raw `Link` types. 

Although `Link`s are generic, the Link Handler, as it currently exists, only provides link references for pages (e.g. `Link<Page>`).
There is no utility in leaving these types as raw in the `LinkHandler`. 

This ambiguity does not create a more flexible API - rather, it forces the user of the service to have an intimate understanding of the inner workings of the link handler implementation in order to divine the reference type and cast it appropriately. 

It also leaves the user of the service vulnerable to implementation changes that could break their application because the API does not convey or make any promises about the reference type; and so it is possible that future changes to the link handler implementation could change the returned reference type and not be an obvious breaking change.

If the `LinkHandler` is to be made public  then this ambiguity should be eliminated.

**Comment by ky940819** [comment]
I have added a PR (#1851) which includes the original PR (#1819) but also addresses the above comment. 
See #1851 for more details.

**Comment by jckautzmann** [comment]
The `LinkHandler` is currently a Sling model and can be reused in following ways:
- adapt from the request: `request.adaptTo(LinkHandler.class)`
- inject into other models using the `@Self` annotation
Doesn't it cover the need described in this ticket?

**Comment by ky940819** [comment]
No, it does not, since `LinkHandler` is not exported from the bundle it is impossible for anyone outside of the bundle to perform `request.adaptTo(LinkHandler.class)`, or to inject it using the `@Self` annotation. 

This issue asks that an interface be created for the `LinkHandler` model, and that the interface be exported by the bundle so that the things you described are possible.

**Comment by dbk-gkrupp** [comment]
Yes, I can only confirm @ky940819 . When trying to inject the LinkHandler I ran into exactly the problem.

**Comment by jkoebner** [comment]
@ky940819 hello there! It is fine for me to close PR #1819 then and do the rest of the discussion on the other PR

**Comment by stefanseifert** [comment]
before making the LinkHandler public, we should discuss if the API is a proper state, created #1861 for this.
